rfc/2019-12-06-dumpling.md

```diff
@@ -0,0 +1,264 @@
+
+## Abstract
+
+We propose introduce a library to replace Mydumper, code named **Dumpling**,
+optimized for TiDB Lightning and to be usable as a library/plugin inside DM and
+TiDB, as well as be an independent program.
+
+## Background
+
+### Optimize output format for machine reading
+
+[Mydumper] is a tool to dump MySQL databases into local filesystem as SQL dump.
+
+[TiDB Lightning] (a data importing tool to TiDB) and [DM] (a platform for
+migration data from MySQL to TiDB) currently relies on Mydumper to extract data
+from an upstream database.
+
+[TiDB Lightning]: https://pingcap.com/docs/dev/reference/tools/tidb-lightning/overview/
+[DM]: https://pingcap.com/docs/dev/reference/tools/data-migration/overview/
+[Mydumper]: https://github.com/pingcap/mydumper/
+
+Currently Lightning needs a [relatively complex tokenizer][t] to parse the SQL
+dump. This slows down the import speed. If we could make the data export tool to
+produce a very easy-to-parse binary output, the processing speed could be
+improved.
+
+[t]: https://github.com/pingcap/tidb-lightning/blob/master/lightning/mydump/parser.rl
+
+However, we do not want to heavily modify Mydumper because of the next reason.
+
+### License and maintenance
+
+Mydumper is licensed in GPLv3. This is incompatible with the license of all
+other PingCAP projects (Apache 2.0). This forces mydumper to only be usable as
+an external program in DM, instead of a library (neither static nor dynamic).
+
+Licensing aside, Mydumper is written in C and GLib, and does not expose itself
+as a library (almost the entire program is included in a single file). This
+further hinders us from tightly integrating it into DM and TiDB, which would
+have expected Go modules.
+
+Finally, the [official Mydumper repository] is sparsely updated, and we prefer
+not to create too much divergence in our fork. These prevent us from investing
+into the Mydumper project, and drive us to create a brand new project.
+
+[official Mydumper repository]: https://github.com/maxbube/mydumper
+
+### New features
+
+As a new project designed for the TiDB ecosystem, we can include further
+features we've desired for long, e.g. writing directly to cloud storage.
+
+## Proposal
+
+We propose introducing a new project **Dumpling**, which is
+
+* Primarily a library written in Go.
+* Support writing SQL dumps from MySQL-compatible database,
+    with speed and concurrency matching Mydumper.
+* Support every feature of Mydumper needed by DM and Lightning, but no more.
+* Support writing in SQL, CSV, and a custom binary format for quick consumption.
+* Support writing to cloud storage besides the local filesystem.
+
+## Rationale
+
+### Name
+
+The initial motivation of this tool is to supplement Lightning.
+We call the new tool "Dumpling" as a portmanteau of "dump" + "Lightning".
+
+### Programming language
+
+We'd like to embed Dumpling into TiDB (as an `EXPORT` statement) and DM
```

Comment:
Once the new physical backup is released, it seems that the only use case for using this against TiDB is to export data to another database (MySQL). In the case of an immediate restore into MySQL the export from TiDB probably won't add convenience because one already has to run an external tool to load the data into MySQL.

The TiDB that runs this might essentially need to be considered offline if the backup process is using up all its resources. So then the value proposition would then be that it is easier to deploy an additional TiDB than to deploy a new tool. This will only be the case if the resource requirements of backup are the same as TiDB. If the resource requirements are bigger, then this won't work. If the resource requirements are smaller and the TiDB node will still serve requests, wrapping as subprocess could still be a good idea to better isolate the backup workload.
Unintelligent load balancing between TiDB could easily lead to the TiDB doing backup to get over-worked.

In contrast, the new physical BR tool ran from inside TiDB would only perform meta operations from TiDB with most of the actual backup work done from TiKV: this should leave most resources still available for the TiDB node.

Reply:
This paragraph explains why we choose Go not other languages. Even if we don't want `EXPORT`, we still need integration with DM.

And given that we're going to have `IMPORT` with Lightning, it is natural to support `EXPORT` as well.

The `EXPORT` statement is not meant to replace BR. BR will be given their own `BACKUP` and `RESTORE` statements.

Reply:
Import with lightning will have some similarly deployment issues since it will use a great deal of CPU. However one of the main use cases is to import when a cluster is first created and no useful queries can be run untill import is complete.

Reply:
True. The `IMPORT` and `EXPORT` statements allow DBAs to manage logical backups via the SQL interface for familiarity. The individual executables are still available though.

Anyway these are getting off-topic.